File: app/src/main/java/com/aayush/filmipedia/model/Genre.java

package com.aayush.filmipedia.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.Expose;

import java.util.Objects;

import androidx.annotation.Nullable;

public class Genre implements Parcelable {
    @Expose
    private Long id;

    @Expose
    private String name;

    private Genre(Parcel in) {
        if (in.readByte() == 0) {
            id = null;
        } else {
            id = in.readLong();
        }
        name = in.readString();
    }

    public static final Creator<Genre> CREATOR = new Creator<Genre>() {
        @Override
        public Genre createFromParcel(Parcel in) {
            return new Genre(in);
        }

        @Override
        public Genre[] newArray(int size) {
            return new Genre[size];
        }
    };

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        if (id == null) {
            dest.writeByte((byte) 0);
        }
        else {
            dest.writeByte((byte) 1);
            dest.writeLong(id);
        }
        dest.writeString(name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj == this) {
            return true;
        }
        if (!(obj instanceof Genre)) {
            return false;
        }

        Genre other = (Genre) obj;

        return Objects.equals(id, other.id) &&
                Objects.equals(name, other.name);
    }
}

File: app/src/main/java/com/aayush/filmipedia/model/Genre.kt
package com.aayush.filmipedia.model

import android.os.Parcel
import android.os.Parcelable
import com.aayush.filmipedia.util.KParcelable
import com.aayush.filmipedia.util.readNullable
import com.aayush.filmipedia.util.writeNullable
import com.google.gson.annotations.Expose

data class Genre(@Expose var id: Long?, @Expose var name: String?): KParcelable {
    private constructor(`in`: Parcel): this(
        `in`.readNullable { `in`.readLong() },
        `in`.readNullable { `in`.readString() }
    )

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeNullable(id, ::writeLong)
        writeNullable(name, ::writeString)
    }

    companion object CREATOR: Parcelable.Creator<Genre> {
        override fun createFromParcel(source: Parcel) = Genre(source)
        override fun newArray(size: Int) = arrayOfNulls<Genre>(size)
    }
}
